Lib/test/test_asyncio/test_streams.py

```diff
@@ -801,6 +801,25 @@ def test_streamreaderprotocol_constructor_use_global_loop(self):
         self.assertEqual(cm.warnings[0].filename, __file__)
         self.assertIs(protocol._loop, self.loop)
 
+    def test_streamreaderprotocol_task_reference(self):
+        # See http://bugs.python.org/issue46309
+        messages = []
+        self.loop.set_exception_handler(lambda loop, ctx: messages.append(ctx))
+
+        async def callback(*args):
+            await self.loop.create_future()
```

Comment:
Callback is leaked, you don't see it because of the custom exception handler.

```suggestion
        future = self.loop.create_future()
        async def callback(*args):
            await future
```

Reply:
Thanks for looking at the changes! I signed the CLA ~3 hours before this PR, so hopefully it will be OK'd sometime today. Your requested changes makes sense. However from my testing, creating the future outside of the callback will not allow the task to be garbage collected. That is, the test will pass even without keeping a reference to the task in `StreamReaderProtocol`.

One solution could be to still create the future inside the callback, but cancel the task via the newly created reference in `StreamReaderProtocol`. Perhaps you have a better suggestion. I can spend some more time on this later today.

Reply:
You are completely correct. I've spent quite a few hours triaging the source of the original issue on stackoverflow and realized the bug is caused when the garbage collector breaks a cyclic reference between the writer, the reader, the task and a future created inside the reader. If any of them are referenced outside of the function, the bug will not appear.

I don't suggest accessing the internals from the test. You're welcome to decline my changes in the tests as long as asyncio doesn't throw any warning about the leak. The garbage collector will take care of it when the loop is closed, or it will be cancelled accordingly.

Reply:
I've accepted your suggested changes in `streams.py`. As for the leaking test callback, perhaps a different solution could be to use `asyncio.all_tasks(self.loop).pop().cancel()` right after the assertion.

Reply:
I believe it's unnecessary, will cause more confusion, and will get bugged if the internals will change and a new task will be created in the background.

Better off letting the cleanup take care of it I believe.